### PR TITLE
Introduce `WeaviateAsyncClient` as async alternative to `WeaviateClient`

### PR DESCRIPTION
**Pull request by tsmith023**
- Changes base implementation to be purely async using `httpx` and `grpc.aio`
- Uses base implementation in `WeaviateAsyncClient` directly
- A class decorator pattern wraps the async implementation in the sync routes replicating the arguments and docstrings automagically @dirkkul
- The API surface itself must be duplicated however through `.pyi` stub fails due to this runtime patching
- Replace `WeaviateClient` internal calls with event-loop-wrapped blocking calls using the base async implementation
- An event loop thread is spun up by the client at instantiation as a global singleton that is used in all calls
- `client.batch` is only available on the `WeaviateClient` object, since it is a purely sync algorithm. It depends on the new higher-level event loop thread

### CODE REVIEW

File: weaviate/collections/batch/batch_wrapper.py [hunk 1]
```diff
@@ -23,6 +30,8 @@ def __init__(self, connection: ConnectionV4, consistency_level: Optional[Consist
 
         self._batch_data = _BatchDataWrapper()
 
+        self._event_loop = event_loop
```
Comment (dirkkul) [inline weaviate/collections/batch/batch_wrapper.py:33]: why are we not using the singleton here, like everywhere else?

File: weaviate/collections/classes/cluster.py [hunk 2]
```diff
@@ -73,7 +73,10 @@ def nodes_verbose(nodes: List[NodeREST]) -> List[Node[Shards, Stats]]:
                         shard_count=node["stats"]["shardCount"],
                     )
                     if "stats" in node
-                    else []
+                    else Stats(
+                        object_count=0,
+                        shard_count=0,
+                    )
```
Comment (dirkkul) [inline weaviate/collections/classes/cluster.py:79]: why this change?

File: weaviate/collections/classes/internal.py [hunk 3]
```diff
@@ -531,6 +527,7 @@ def from_input(
     ObjectSingleReturn[Properties, TReferences],
     ObjectSingleReturn[TProperties, References],
     ObjectSingleReturn[TProperties, CrossReferences],
+    None,
```
Comment (dirkkul) [inline weaviate/collections/classes/internal.py:530]: what is this for?

File: weaviate/collections/collection/async_.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+            `weaviate.UnexpectedStatusCodeError`
+                If weaviate reports a none OK status.
+            `weaviate.EmptyResponseError`
+                If the response is empty.
+        """
+        return [
+            shard
+            for node in await self.__cluster.nodes(self.name, output="verbose")
+            for shard in node.shards
+        ]
```
Comment (dirkkul) [inline weaviate/collections/collection/async_.py:350]: can we share some code between them?

File: weaviate/collections/collection/sync.py [hunk 5]
```diff
@@ -63,58 +66,83 @@ def __init__(
...
         """This namespace includes all the CUD methods available to you when modifying the data of the collection in Weaviate."""
-        self.generate = _GenerateCollection(
-            connection,
-            self.name,
-            consistency_level,
-            tenant,
-            properties,
-            references,
-            validate_arguments,
+        self.generate = _GenerateCollection[Properties, References](
```
Comment (dirkkul) [inline weaviate/collections/collection/sync.py:118]: there are also some unrelated changes in here?

File: weaviate/collections/tenants/async_.py [hunk 6]
```diff
@@ -289,16 +275,21 @@ def exists(self, tenant: Union[str, Tenant]) -> bool:
...
-                ok_in=[200, 404], error=f"Check if tenant exists for {self.__name}"
+                ok_in=[200, 404], error=f"Check if tenant exists for {self._name}"
             ),  # allow 404 to perform bool check on response code
         )
         return response.status_code == 200
+
+
+@asyncify.convert
+class _Tenants(_TenantsAsync):
+    pass
```
Comment (dirkkul) [inline weaviate/collections/tenants/async_.py:295]: seems to be a duplicate?

File: weaviate/event_loop.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        in an event loop running in a parallel thread.
+
+        The result of the coroutine is returned, either when the coroutine completes or raises an exception.
+        """
+        if self.loop is None or self.loop.is_closed():
+            raise WeaviateClosedClientError()
+        fut = asyncio.run_coroutine_threadsafe(f(*args, **kwargs), self.loop)
+        return fut.result()
+
+    def schedule(
```
Comment (dirkkul) [inline weaviate/event_loop.py:42]: does not seem to be used?

File: weaviate/embedded.py [hunk 8]
```diff
@@ -199,20 +202,23 @@ def ensure_running(self) -> None:
...
         self.ensure_weaviate_binary_exists()
         my_env = os.environ.copy()
 
         my_env.setdefault("AUTHENTICATION_ANONYMOUS_ACCESS_ENABLED", "true")
         my_env.setdefault("QUERY_DEFAULTS_LIMIT", "20")
         my_env.setdefault("PERSISTENCE_DATA_PATH", self.options.persistence_data_path)
         # Bug with weaviate requires setting gossip and data bind port
-        my_env.setdefault("CLUSTER_GOSSIP_BIND_PORT", str(get_random_port()))
+        my_env.setdefault("CLUSTER_JOIN", f"localhost:{random_ports[3]}")
+        my_env.setdefault("CLUSTER_GOSSIP_BIND_PORT", str(random_ports[3]))
```
Comment (dirkkul) [inline weaviate/embedded.py:215]: these should probably not be the same? But this should also not be part of this PR?

File: weaviate/connect/v4.py [hunk 9]
```diff
@@ -93,25 +97,24 @@ def __init__(
         trust_env: bool,
         additional_headers: Optional[Dict[str, Any]],
         connection_config: ConnectionConfig,
+        loop: asyncio.AbstractEventLoop,  # required for background token refresh
```
Comment (dirkkul) [inline weaviate/connect/v4.py:100]: singleton?

**Review by dirkkul — CHANGES_REQUESTED** [review]

File: integration/test_collection_aggregate.py [hunk 10]
```diff
@@ -51,15 +54,11 @@ def test_collection_length_tenant(collection_factory: CollectionFactory, how_man
...
-        tenants=[Tenant(name="tenant1"), Tenant(name="tenant2"), Tenant(name="tenant3")]
-    )
+    collection.tenants.create(tenants=[Tenant(name="tenant1"), Tenant(name="tenant2")])
     collection.with_tenant("tenant1").data.insert_many([{} for _ in range(how_many)])
-    collection.with_tenant("tenant2").data.insert_many([{} for _ in range(how_many * 2)])
 
-    assert len(collection.with_tenant("tenant2")) == how_many * 2
-    assert len(collection.with_tenant("tenant3")) == 0
+    assert len(collection.with_tenant("tenant1")) == how_many
+    assert len(collection.with_tenant("tenant2")) == 0
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply integration/test_collection_aggregate.py:58]: I remember these changes being part of the debugging process for some of the flakes. I'll undo them since it wasn't this in the end!

File: weaviate/collections/batch/batch_wrapper.py [hunk 11]
```diff
@@ -44,32 +53,35 @@ def wait_for_vector_indexing(
...
                 )
                 if how_many_failures == how_many:
                     raise e
                 time.sleep(2**how_many)
-                return is_ready(how_many + 1)
+                return await is_ready(how_many + 1)
 
         count = 0
-        while not is_ready(0):
+        while not self._event_loop.run_until_complete(is_ready, count):
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/batch/batch_wrapper.py:75]: Because `ConnectionV4` only contains `async` methods now using `httpx.AsyncClient` and `grpclib.Channel` so rather than persist a `sync` version of `ConnectionV4`, all the functionality above it has to use the root `async` implementation either directly in an `async` method or wrapped by the `_EventLoop` code

An added benefit here of having `async def is_ready` is that the `__get_shards_readiness` calls can be ran concurrently, using `asyncio.gather`, rather than serially as it was before

File: weaviate/client.py [hunk 12]
```diff
@@ -517,3 +725,77 @@ def __del__(self) -> None:
         # in case an exception happens before definition of the client
         if hasattr(self, "_connection"):
             self._connection.close()
+
+    def is_ready(self) -> bool:
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/client.py:730]: I'll try and move as much of the shared functionality into a base client again

File: integration/test_client.py [hunk 13]
```diff
@@ -323,47 +321,53 @@ def test_collection_name_capitalization(
 
 
 def test_client_cluster(client: weaviate.WeaviateClient, request: SubRequest) -> None:
-    client.collections.delete(request.node.name)
-    collection = client.collections.create(
-        name=request.node.name, vectorizer_config=Configure.Vectorizer.none()
-    )
+    try:
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply integration/test_client.py:324]: These changes are so that collections are cleaned up appropriately if the tests fail, they were part of the debugging process of the testing flakes due to the implementation

File: integration/test_iterator.py [hunk 14]
```diff
@@ -196,7 +196,7 @@ def test_iterator(collection_factory: CollectionFactory, count: int) -> None:
     # make sure a new iterator resets the internal state and that the return order is the same for every run
     for _ in range(3):
         # get the property and sort them - order returned by weaviate is not identical to the order inserted
-        ret: list[int] = [obj.properties["data"] for obj in collection.iterator()]
+        ret: list[int] = [int(obj.properties["data"]) for obj in collection.iterator()]
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply integration/test_iterator.py:199]: This was to fix a mypy/pyright issue that was suddenly flagged due to a change in their deps on `main`

File: integration_embedded/test_client.py [hunk 15]
```diff
@@ -50,29 +60,49 @@ def test_embedded_as_context_manager() -> None:
...
+    client = weaviate.connect_to_embedded(
+        port=8099, grpc_port=50150, environment_variables={"DISABLE_TELEMETRY": "true"}
+    )
+    try:
+        with pytest.raises(weaviate.exceptions.WeaviateStartUpError):
+            weaviate.connect_to_embedded(
+                port=8100, grpc_port=50150, environment_variables={"DISABLE_TELEMETRY": "true"}
+            )
+    finally:
+        client.close()
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply integration_embedded/test_client.py:108]: Same as the clauses for deleting collections. In this case, `grpclib` throws warnings if its `Channel` is not closed when cleaned up by the GC

File: profiling/test_profiling.py [hunk 16]
```diff
@@ -132,7 +135,7 @@ def test_object_by_id(client: weaviate.WeaviateClient) -> None:
 
     batchReturn = col.data.insert_many([{"index": i} for i in range(1000)])
 
-    for i in range(1000):
+    for i in tqdm.trange(1000):
```
[the remark this answers is not in the document]
Comment (dirkkul) [reply profiling/test_profiling.py]: I made some comments the last time, but didn't submit them - must have been a leftover

File: weaviate/collections/aggregations/base.py [hunk 17]
```diff
@@ -36,15 +38,18 @@
 from weaviate.collections.filters import _FilterToREST
 from weaviate.exceptions import WeaviateInvalidInputError, WeaviateQueryError
 from weaviate.gql.aggregate import AggregateBuilder
-from weaviate.util import file_encoder_b64
+from weaviate.util import file_encoder_b64, _decode_json_response_dict
 from weaviate.validator import _ValidateArgument, _validate_input
 from weaviate.types import NUMBER, UUID
 
+if TYPE_CHECKING:
+    pass
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/aggregations/base.py:46]: Indeed, it does nothing!

File: weaviate/collections/backups/base.py [hunk 18]
```diff
@@ -0,0 +1,13 @@
+from weaviate.backup.backup import (
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/backups/base.py:1]: Would you prefer something like `weaviate/collections/backups/backups.py`?

File: weaviate/collections/batch/base.py [hunk 19]
```diff
@@ -284,69 +257,54 @@ def _shutdown(self) -> None:
...
+                    objs,
+                    refs,
+                    readd_rate_limit=isinstance(self.__batching_mode, _RateLimitedBatching),
+                )
 
-        finally:
-            future = asyncio.run_coroutine_threadsafe(self.__connection.aclose(), loop)
-            future.result()  # Wait for self._connection.aclose() to finish
-            loop.call_soon_threadsafe(loop.stop)
+            time.sleep(refresh_time)
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/batch/base.py:307]: The event loop is now managed at the client level instead of within each batch. When a batching job is created, it just fetches the current event loop without spawning its own. Therefore, connection opening and closing is handled at the client-level in the `client.connect()` method and so no longer occurs in the batch init processes

File: weaviate/collections/batch/base.py [hunk 20]
```diff
@@ -709,3 +667,23 @@ def __check_bg_thread_alive(self) -> None:
             return
 
         raise self.__bg_thread_exception or Exception("Batch thread died unexpectedly")
+
+
+class _ClusterBatch:
+    def __init__(self, connection: ConnectionV4):
+        self._connection = connection
+
+    async def get_nodes_status(
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/batch/base.py:676]: Just as with the methods in `batch_wrapper.py`, the fundamental connection is `async` only so we need to have async methods wrapping these calls that are scheduled on the event loop

File: weaviate/collections/batch/batch_wrapper.py [hunk 21]
```diff
@@ -79,6 +91,9 @@ def _get_shards_readiness(self, shard: Shard) -> List[bool]:
             for shard in res
         ]
 
+    def _get_shards_readiness(self, shard: Shard) -> List[bool]:
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/batch/batch_wrapper.py:94]: I felt that it was cleaner to have the singleton called only once here and injected where it was needed. In the case of the decorator, that's the only place the singleton is called then the decorator is added wherever. In this case, I'd like the same logic in that we call the singleton once and then apply it everywhere it needs to go

File: weaviate/collections/batch/batch_wrapper.py (hunk 1 above)
Comment (tsmith023) [reply weaviate/collections/batch/batch_wrapper.py:33]: I felt that it was cleaner to have the singleton called only once here and injected where it was needed. In the case of the decorator, that's the only place the singleton is called then the decorator is added wherever. In this case, I'd like the same logic in that we call the singleton once and then apply it everywhere it needs to go

File: weaviate/collections/batch/batch_wrapper.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/batch/batch_wrapper.py:94]: Same as above, fundamental `ConnectionV4` implementation is `async` now

File: weaviate/collections/classes/cluster.py (hunk 2 above)
Comment (tsmith023) [reply weaviate/collections/classes/cluster.py:79]: `pyright` didn't like `[]` being returned where `Stats` is expected. The previous version before the change on `main` must have missed it

File: weaviate/collections/classes/internal.py (hunk 3 above)
Comment (tsmith023) [reply weaviate/collections/classes/internal.py:530]: The `fetch_object_by_id` query returns `None` if `len(objects.objects) == 0` but this wasn't captured in the type `QuerySingleReturn` so I've added it here for that

File: weaviate/collections/collection/async_.py (hunk 4 above)
Comment (tsmith023) [reply weaviate/collections/collection/async_.py:350]: Yeh I'll try to achieve some of that

File: weaviate/collections/collection/sync.py (hunk 5 above)
Comment (tsmith023) [reply weaviate/collections/collection/sync.py:118]: This is to fix a typing issue raised by `pyright` that was likely missed in previous versions

File: weaviate/event_loop.py (hunk 7 above)
Comment (tsmith023) [reply weaviate/event_loop.py:42]: It's used on `L300` of `weaviate.collections.batch.base.py`

File: weaviate/embedded.py (hunk 8 above)
Comment (tsmith023) [reply weaviate/embedded.py:215]: Ahh, another part of debugging that I did not undo 😅 Will remove!

File: weaviate/connect/v4.py (hunk 9 above)
Comment (tsmith023) [reply weaviate/connect/v4.py:100]: This needs to be just an asyncio event loop in the case that they are running with `WeaviateAsyncClient` in which case they need to pass the running event loop to the periodic refresh token background thread

File: weaviate/collections/aggregations/near_text.pyi [hunk 22]
```diff
@@ -0,0 +1,112 @@
+from typing import List, Literal, Optional, Union, overload
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/aggregations/near_text.pyi:1]: Yep, will add some more async tests

File: weaviate/collections/backups/async_.py [hunk 23]
```diff
@@ -1,23 +1,18 @@
 from typing import Optional
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/backups/async_.py:1]: I can't think of a better convention than describing the files by what they do. Do you not think the directory structure is self-explanatory?

File: weaviate/collections/backups/base.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (dirkkul) [reply weaviate/collections/backups/base.py:1]: I close this, let's discuss above

File: weaviate/collections/batch/base.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (dirkkul) [reply weaviate/collections/batch/base.py:307]: ah sorry - you mentioned there was a bug in the batch loop

File: weaviate/collections/backups/async_.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/backups/async_.py:1]: Oh interesting! VS code has the dir structure as a breadcrumb, which is certainly why I went with this approach. I'm happy to go with your suggestion :)

File: weaviate/collections/batch/base.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/batch/base.py:307]: Ah yeh, also that!

File: weaviate/collections/backups/async_.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (dirkkul) [reply weaviate/collections/backups/async_.py:1]: > I'm happy to go with your suggestion :)

great :)

> VS code has the dir structure as a breadcrumb

Pycharm too, but only if you have more than one file open.

File: weaviate/collections/batch/base.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (dirkkul) [reply weaviate/collections/batch/base.py:307]: What was the bug? :)

File: weaviate/collections/batch/base.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (dirkkul) [reply weaviate/collections/batch/base.py:676]: Why not just use ClusterSync then?

File: weaviate/event_loop.py (hunk 7 above)
Comment (dirkkul) [reply weaviate/event_loop.py:42]: the github UI didn't show that :)

Comment (tsmith023) [reply weaviate/event_loop.py:42]: Probably overloaded with the number of changed lines 😅

File: weaviate/client.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/client.py:342]: Have merged some of the shared functionality and used `syncify.convert` to automate the implementation. It leads to redefinition of the interface in `client.pyi`, however, like in other areas of the codebase. So there's at tradeoff!

File: weaviate/collections/batch/base.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/batch/base.py:307]: There was a timing issue in the rate limiting workflow due to some of the logic. Due to the time taken to spin up the event loop, it wasn't apparent. However, with the event loop managed elsewhere, this delay doesn't occur and the error surfaces. I'll check again if it's still present

File: weaviate/collections/batch/batch_wrapper.py (hunk 1 above)
Comment (dirkkul) [reply weaviate/collections/batch/batch_wrapper.py]: I'm not 100% sure if I can follow 😅 would you mind expanding a bit?

File: weaviate/collections/batch/base.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (dirkkul) [reply weaviate/collections/batch/base.py:307]: great, let's resolve this after you confirmend it is gone!

File: weaviate/collections/batch/batch_wrapper.py (hunk 1 above)
Comment (tsmith023) [reply weaviate/collections/batch/batch_wrapper.py]: I've revisited this code and you're right, it's more confusing prop drilling here and not in other places. Have changed in the latest version!

File: weaviate/connect/v4.py (hunk 9 above)
Comment (dirkkul) [reply weaviate/connect/v4.py:100]: same here, don't fully understand the reasoning

Comment (tsmith023) [reply weaviate/connect/v4.py:100]: The `_ConnectionV4` class is the fundamental implementation of the client's connection to Weaviate in both the `async` and `sync` use-cases. In both cases, the background token refreshing logic requires an open and running `asyncio.AbstractEventLoop` in order to schedule the defined coroutines on.

In the case of the `sync` parts of the codebase, this loop is managed by the `_EventLoop` class that is retrieved using the `_EventLoopSingleton` factory class. In the `async` parts of the codebase this loop already exists as the `asyncio.get_running_event_loop()`, which is called at `weaviate/client.py#146` and injected at `weaviate/client_base.py#84`.

As such, it makes sense to me that `_ConnectionV4` should depend on the `asyncio` event loop rather than on the singleton since the singleton is only relevant to the `sync` methods, which sit above `_ConnectionV4` in the structure of the codebase.

Comment (dirkkul) [reply weaviate/connect/v4.py:100]: thank you!

File: weaviate/collections/batch/base.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/batch/base.py:676]: Hmm, in order to use `_ClusterSync` there needs to be a refactoring of the batch logic and/or the logic in `_ClusterSync` to handle the potential returns of Weaviate depending on the present of async indexing. This should be changed and fixed but I think it should go in another PR (to reduce this one's size and scope). WDYT?

[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/batch/base.py:676]: For example, `batch_stats` isn't a field in `Node` in `weaviate.collections.classes.cluster` when it should be but its presence depends on async indexing so needs to be handled appropriately

[the remark this answers is not in the document]
Comment (dirkkul) [reply weaviate/collections/batch/base.py:676]: sounds good - want to create an issue so we do not forget?

File: weaviate/collections/batch/base.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/batch/base.py:307]: Have added tests for rate limiting on large and small ingestion sizes and there are no issues :D

**Comment by tsmith023** [comment]
@rolandgvc, we have released `v4.7.0b0` as an initial pre-release of the new async implementation!

We're aiming to have a week or two of community engagement on the _beta_ to get feedback and find bugs. If you're happy to test it out it would be great to hear your feedback too 😁

**Comment by tsmith023** [comment]
Here's some example usage to see how it works!
```python
import weaviate
import weaviate.classes as wvc

async def async_example() -> None:
    async with weaviate.use_async_with_local() as client:
        collection = await client.collections.create(
            properties=[
                wvc.config.Property(name="name", data_type=DataType.TEXT),
            ],
            vectorizer_config=wvc.config.Configure.Vectorizer.none(),
        )
        await collection.data.insert_many(
            [
                {"name": "John Doe"},
            ]
        )
        res = await collection.query.fetch_objects()
        assert len(res.objects) == 1
        assert res.objects[0].properties["name"] == "John Doe"
```

**Comment by gokturkDev** [comment]
Hi great work y'all we really needed this 👏 Is there some documentation about this? From what I gather in test cases, we can use use_async flag on client instantiation and sync functions become async?

**Comment by gokturkDev** [comment]
One more, is this ready for production code yet or should we wait for the full release?

**Comment by dirkkul** [comment]
> One more, is this ready for production code yet or should we wait for the full release?

please wait for the full release for production - but please try it out now so we can release it as quickly as possible :)

> Is there some documentation about this? From what I gather in test cases, we can use use_async flag on client instantiation and sync functions become async?

Not yet - basically you instantiate the async client and then all functions are async

**Comment by gokturkDev** [comment]
Great looking forward to the release 👍  This might be too big of a hussle now but I would suggest embedding the async functions into WeaviateClient rather than creating a separate AsyncClient class, as a lot of projects (and ours) use a wrapper class for interacting with weaviate, therefore a separate async wrapper class is needed instead of just adding new async methods, which is a bit inconvenient

**Comment by dirkkul** [comment]
>  This might be too big of a hussle now but I would suggest embedding the async functions into WeaviateClient rather than creating a separate AsyncClient class, as a lot of projects (and ours) use a wrapper class for interacting with weaviate, therefore a separate async wrapper class is needed instead of just adding new async methods, which is a bit inconvenient

The sync class is now basically a wrapper around the async functions - the clients itself are duplicated for typing reasons, but almost everything else is async under the hood and called from the sync client

File: integration/test_collection_config.py [hunk 24]
```diff
@@ -942,14 +942,15 @@ def test_config_unknown_module(request: SubRequest) -> None:
 
 def test_create_custom_module(collection_factory: CollectionFactory) -> None:
     collection = collection_factory(
+        name="one",
```
Comment (dirkkul) [inline integration/test_collection_config.py:945]: maybe use a global counter instead in the conftest to add to the collection names? We have the GIL, so we wont have real races

so something like
```
name_fixture = _sanitize_collection_name(request.node.name) + name + "_" +counter
counter ++
```

Comment (tsmith023) [reply integration/test_collection_config.py:945]: Hmm, not sure I understand. The tests [were all failing](https://github.com/weaviate/weaviate-python-client/actions/runs/9775672600/job/26986616255?pr=1007) with a collection-already-exists error so there is a race happening in the test. My guess is that the GC isn't cleaning up `collection` before `collection_2` gets created now, for whatever reason, so the names need to be different

I will add the counter though, much better solution

Comment (dirkkul) [reply integration/test_collection_config.py:945]: sorry, I moved some words around and then it didn't make sense anymore.

basically just wanted to automate that the names are unique when creating from the same test :)

**Review by dirkkul — APPROVED** [review]
I think we tested enough :)

**Comment by gokturkDev** [comment]
Hi when is this going to be in the release version
